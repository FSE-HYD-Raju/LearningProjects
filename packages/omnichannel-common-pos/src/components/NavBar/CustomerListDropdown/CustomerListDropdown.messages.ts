/* tslint:disable:max-line-length */
import { defineMessages, FormattedMessage } from "react-intl";

// Interface has been generated automatically. Please do not edit manually.
// For changes run in console: npm run rebuild-messages-interfaces
interface CustomerListDropdownMessagesType {
	createCustomer: FormattedMessage.MessageDescriptor;
	customerAddress: FormattedMessage.MessageDescriptor;
	customerName: FormattedMessage.MessageDescriptor;
	customerNumber: FormattedMessage.MessageDescriptor;
	hyphenSymbol: FormattedMessage.MessageDescriptor;
	items: FormattedMessage.MessageDescriptor;
	noCustomerAccount: FormattedMessage.MessageDescriptor;
	noResultsFound: FormattedMessage.MessageDescriptor;
	ofText: FormattedMessage.MessageDescriptor;
	searchPlaceholder: FormattedMessage.MessageDescriptor;
	searching: FormattedMessage.MessageDescriptor;
}
const CustomerListDropdownMessages: CustomerListDropdownMessagesType = defineMessages({
	createCustomer: {
		id: "Customer-list-dropdown-create-customer-button-label",
		description: "Create customer button in customer list",
		defaultMessage: "Create customer"
	},
	customerAddress: {
		id: "Customer-list-dropdown-address",
		description: "Address",
		defaultMessage: "Address"
	},
	customerName: {
		id: "Customer-list-dropdown-name",
		description: "Name",
		defaultMessage: "Name"
	},
	customerNumber: {
		id: "Customer-list-dropdown-phone-number",
		description: "Phone number",
		defaultMessage: "Number"
	},
	hyphenSymbol: {
		id: "Customer-list-dropdown-pagination-hyphen-symbol",
		description: "pagination info",
		defaultMessage: "-"
	},
	items: {
		id: "Customer-list-dropdown-pagination-items",
		description: "pagination info",
		defaultMessage: "items"
	},
	noCustomerAccount: {
		id: "Customer-list-dropdown-no-customer-account",
		description: "No customer accounts exists ",
		defaultMessage: "No customer account. View individual details."
	},
	noResultsFound: {
		id: "Customer-list-dropdown-could-not-find-customer-with",
		description: "Message shown when no results were found in search",
		defaultMessage: "Couldn't find anything with {terms}"
	},
	ofText: {
		id: "Customer-list-dropdown-pagination-of-text",
		description: "pagination info",
		defaultMessage: "of"
	},
	searchPlaceholder: {
		id: "Customer-list-dropdown-placeholder",
		description: "Search Placeholder",
		defaultMessage: "Filter results"
	},
	searching: {
		id: "Customer-list-dropdown-searching-for-customer",
		description: "Message shown when customer search in is in progress",
		defaultMessage: "Searching for {terms}"
	},
});

export default CustomerListDropdownMessages;
export { CustomerListDropdownMessages, CustomerListDropdownMessagesType };
