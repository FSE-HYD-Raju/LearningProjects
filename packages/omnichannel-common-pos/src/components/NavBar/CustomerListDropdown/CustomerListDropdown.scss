.CustomerListDropdown {
	&-search-wrapper {
		margin-bottom: 1rem;
		.fas {
			padding-right: 0.25rem;
		}
		div {
			padding-bottom: 0.1rem;
		}
	}
	&-name {
		margin: 1rem 0 1rem 0;
	}

	&-customer-list-footer {
		border-top: 1px solid #e2e2e2;
	}
	&-account-type {
		color: gray("500");
		text-transform: capitalize;
	}
	&-dropdown-wrapper {
		padding: 10px 10px 0 10px;
	}
	&-tile-wrapper {
		border-top: 1px solid #e2e2e2;
	}
	&-pagination-buttons {
		display: inline-block;
		padding-left: 0.8rem;
	}
	&-pagination-container {
		display: inline-block;
	}

	&-accounts-wrapper {
		display: flex;
		justify-content: space-between;
		margin-bottom: 0.5rem;
	}
	&-search {
		max-width: 50%;
		display: inline-block;
	}
	&-customer-dropdown-header {
		display: flex;
		padding-bottom: 0.2rem;
		justify-content: space-between;
	}
	&-link {
		color: $main-blue;
		cursor: pointer;
	}
	&-no-account {
		padding-bottom: 0.2rem;
	}

	@include media-breakpoint-up($b2c-desktop-nav-breakpoint) {
		&-customer-list-dropdown.dropdown-menu {
			width: 550px;
		}
		&-customer-list-searching,
		&-customer-list-nothing-found {
			font-size: $font-size-lg;
			padding: $table-cell-padding;
			color: gray("500");
			margin: 0;
		}
		&-customer-list-table {
			margin-bottom: 0;
			white-space: nowrap;
		}
		&-customer-list-footer-button {
			display: block;
			padding: $table-cell-padding;
		}
	}
}
