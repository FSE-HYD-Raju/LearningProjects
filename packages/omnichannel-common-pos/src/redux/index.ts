"use strict";

import { default as actions, AppActions } from "./actions";
import { default as reducers, AppState } from "./reducers";
import { InterceptReducers } from "./reducers/reducerUtils";

import { default as DevTools } from "./DevTools";

import { connectFluxRedux, getDefaultFluxStoreMap } from "./connectFluxRedux";
import { configureStore, AppStore, StoreConfig } from "./store";
import * as actionTypes from "./actions";
import * as decorators from "./decorators";

import * as channelUtils from "./utils/channel";
export * from "./utils";
import * as Selectors from "./selectors";

import { initialState } from "./reducers/initialState";
import BaseService from "./services/BaseService";
export * from "./settings/core";
import { productToBasketItem as productToBasketItemTypes } from "./models/feature/feature.types";
// uses "configureStore" from ./store, should be imported after ./store
import * as ReduxTestUtils from "./utils/ReduxTestUtils";

export {
	decorators,
	Selectors,
	actions,
	reducers,
	DevTools,
	StoreConfig,
	InterceptReducers,
	AppStore,
	initialState,
	AppActions,
	AppState,
	connectFluxRedux,
	configureStore,
	getDefaultFluxStoreMap,
	ReduxTestUtils,
	channelUtils,
	actionTypes,
	BaseService,
	productToBasketItemTypes,
};

export * from "./actions/actionsPayload";
export * from "./services/Rest";
export * from "./services";
export * from "./types";
export * from "./decorators";
export * from "./models/msisdnSelection/msisdnSelection.actions";
export * from "./models/productOfferingConfiguration/productOfferingConfiguration.utils";
