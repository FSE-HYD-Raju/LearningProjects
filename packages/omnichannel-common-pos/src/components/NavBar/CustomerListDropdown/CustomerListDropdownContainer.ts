import { Dispatch, AnyAction } from "redux";
import { connect } from "react-redux";
import { CustomerAccount } from "../../../redux";
import CustomerListDropdown, { CustomerListDropdownActionProps } from "./CustomerListDropdown";
import actions from "../../../redux/actions";

const mapDispatchToProps = (dispatch: Dispatch<AnyAction>): CustomerListDropdownActionProps => {
	return {
		actions: {
			setActiveCustomerAccount: (customerAccount: CustomerAccount) => {
				dispatch(actions.customer.setActiveCustomerAccount(customerAccount));
			},
		}
	};
};

export default connect(null, mapDispatchToProps)(CustomerListDropdown);
