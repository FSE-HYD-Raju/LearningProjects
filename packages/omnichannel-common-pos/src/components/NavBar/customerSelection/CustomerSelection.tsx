import CustomerListDropdownContainer from "../CustomerListDropdown/CustomerListDropdownContainer";
import cssns from "../../../utils/cssnsConfig";
import { PureComponent } from "react";
import { RouteComponentProps, withRouter, StaticContext } from "react-router";
import { isEmpty } from "lodash";
import { User, Customer, CustomerAccount } from "../../../redux";
import NavBarSearchMenu from "../NavBarSearchMenu/NavBarSearchMenu";
import { commonServiceDeskRoutes } from "../../../routes/commonRoutesMap";
import IntlContainer from "../../../channelUtils/IntlContainer";
import { OcButton, OcButtonType } from "../../ocComponents";

const { React } = cssns("NavBar");

interface CustomerSelectionState {
	criteria: string;
	showCustomerListDropdown: boolean;
	confirmModalVisible: boolean;
}

interface CustomerSelectionStateProps {
	user?: User;
	searchConfigs: string;
	singleTerm?: string;
	singleTermCustomers?: Array<Customer>;
	searchingForSingleCustomer: boolean;
	hasActiveCustomerCase: boolean; // !!activeCusterCase
}

interface CustomerSelectionActionProps {
	actions: {
		clearSingleTermSearch: () => void;
		searchCustomerWithSingleTerm: (term: string, criteria: string) => void;
		setSingleTerm: (ter: string) => void;
		showCustomerCreationModal: (show: boolean) => void;
		setCustomer: (customer: Customer, active: boolean) => void;
		createNewCustomerCase: (salesRepId?: string) => void;
		deleteUIbasket: () => void;
		cancelAddProduct: () => void;
		resetConfigurations: () => void;
		resetAddressWithBasketItemIdEntries: () => void;
		resetLocations: () => void;
	};
}

type CustomerSelectionProps = CustomerSelectionStateProps & CustomerSelectionActionProps & RouteComponentProps<any>;

class CustomerSelection extends PureComponent<CustomerSelectionProps, CustomerSelectionState> {
	static displayName: string = "CustomerSelection";

	constructor(props: CustomerSelectionProps) {
		super(props);

		this.state = {
			criteria: "name",
			showCustomerListDropdown: false,
			confirmModalVisible: false
		};
	}

	searchCustomers = (e: React.SyntheticEvent<any>) => {
		e.preventDefault();

		if (!this.props.singleTerm || isEmpty(this.props.singleTerm) || isEmpty(this.props.singleTerm.trim())) {
			this.setState({
				showCustomerListDropdown: false
			});
			this.props.actions.clearSingleTermSearch();
			return;
		}

		this.props.actions.searchCustomerWithSingleTerm(this.props.singleTerm, this.state.criteria);

		this.setState({
			showCustomerListDropdown: true
		});
	};

	handleCriteriaChange = (val: string) => {
		this.setState({
			criteria: val
		});
	};

	handleSearchInput = (e: React.ChangeEvent<HTMLInputElement>) => {
		this.props.actions.setSingleTerm(e.target.value);
	};

	closeCustomerSearch = (resetTerm?: boolean) => {
		if (resetTerm) {
			this.props.actions.setSingleTerm("");
		}

		this.setState({
			showCustomerListDropdown: false
		});
	};

	selectCustomer = (customer: Customer) => {
		if (this.state.showCustomerListDropdown) {
			this.props.actions.deleteUIbasket();
			this.props.actions.cancelAddProduct();
			this.props.actions.resetConfigurations();
			this.props.actions.createNewCustomerCase();
			this.props.actions.resetAddressWithBasketItemIdEntries();
			this.props.actions.resetLocations();

		}
		this.props.actions.setCustomer(customer, true);
		this.props.actions.setSingleTerm("");
		this.closeCustomerSearch(true);
		this.props.history.push(commonServiceDeskRoutes.SERVICE_DESK_CUSTOMER.createLink());
	};

	showCustomerCreationModal = (show: boolean) => {
		this.closeCustomerSearch();
		this.props.actions.showCustomerCreationModal(show);
	};

	render() {
		if (!this.props.hasActiveCustomerCase) {
			return null;
		}

		return (
			<form className="customer-selection" id="w-app-search" onSubmit={this.searchCustomers}>
				{this.state.showCustomerListDropdown && (
					<CustomerListDropdownContainer
						singleTermCustomers={this.props.singleTermCustomers}
						selectCustomer={this.selectCustomer}
						closeCustomerSearch={this.closeCustomerSearch}
						showCustomerCreationModal={() => this.showCustomerCreationModal(true)}
						searchTerm={this.props.singleTerm}
						searchingForSingleCustomer={this.props.searchingForSingleCustomer}
					/>
				)}
				<div className="input-group">
					<IntlContainer>
						<NavBarSearchMenu
							criteria={this.state.criteria}
							unparsedSearchConfigs={this.props.searchConfigs}
							handleCriteriaChange={this.handleCriteriaChange}
						/>
					</IntlContainer>
					<input
						id="navbar-search"
						className="form-control"
						onChange={this.handleSearchInput}
						value={this.props.singleTerm}
					/>
					<div className="input-group-append">
						<OcButton
							htmlBtnType="submit"
							buttonType={OcButtonType.DEFAULT}
							id="navbar_search_button"
							onClick={this.searchCustomers}
						>
							<i className="fa fa-search" />
						</OcButton>
					</div>
				</div>
			</form>
		);
	}
}

const CustomerSelectionWithRouter: React.ComponentClass<
	CustomerSelectionStateProps & CustomerSelectionActionProps
> = withRouter<CustomerSelectionProps>(CustomerSelection);

export default CustomerSelectionWithRouter;

export {
	RouteComponentProps,
	StaticContext,
	CustomerSelectionActionProps,
	CustomerSelectionStateProps,
	CustomerSelectionProps,
	CustomerSelectionState,
	CustomerSelection as CustomerSelectionUnwrapped
};
