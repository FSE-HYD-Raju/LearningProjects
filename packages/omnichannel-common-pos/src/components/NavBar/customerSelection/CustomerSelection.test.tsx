import * as React from "react";
import { CustomerSelectionProps, CustomerSelectionUnwrapped as CustomerSelection } from "./CustomerSelection";
import { User } from "../../../redux";
import { RouteComponentProps } from "react-router";
import { ReactWrapper } from "enzyme";
import { CustomerListDropdownProps } from "../CustomerListDropdown/CustomerListDropdown";
import {
	mountWithContext,
	shallowWithContext,
	TestUtils,
	SimpleDataMock,
	mountWithContextAndRouterProps
} from "../../../testUtils";
const { makeStore } = TestUtils;

describe("CustomerSelection", () => {
	let minProps: CustomerSelectionProps;
	let user: User;
	let context: any;
	let store: any;
	let consulStore: any;

	beforeEach(() => {
		consulStore = {
			searchConfigs:
				"{\"POSSearchConfigs\":[{\"id\": \"name\", \"enabled\": true}," +
				"{\"id\": \"phone-number\",\"enabled\": true}," +
				"{\"id\": \"passport\",\"identificationType\": \"passport\",\"enabled\": true}," +
				"{\"id\": \"minor-ID\",\"identificationType\": \"minor-ID\",\"enabled\": true}," +
				"{\"id\": \"ID-card\",\"identificationType\": \"identification-card\",\"enabled\": true}," +
				"{\"id\": \"military-ID\",\"identificationType\": \"military-ID\",\"enabled\": true}," +
				"{\"id\": \"foreign-document\",\"identificationType\": \"foreign-document\",\"enabled\": true}," +
				"{\"id\": \"diplomatic-carnet\",\"identificationType\": \"diplomatic-carnet\",\"enabled\": true}," +
				"{\"id\": \"student-ID\",\"identificationType\": \"student-ID\",\"enabled\": true}," +
				"{\"id\": \"tax-ID\",\"identificationType\": \"tax-ID\",\"enabled\": true}," +
				"{\"id\": \"drivers-licence\",\"identificationType\": \"drivers-licence\",\"enabled\": true}," +
				"{\"id\": \"residence-number\",\"identificationType\": \"residence-number\",\"enabled\": true}," +
				"{\"id\": \"work-permit\",\"identificationType\": \"work-permit\",\"enabled\": true}," +
				"{\"id\": \"disability-cell\",\"identificationType\": \"disability-cell\",\"enabled\": true}," +
				"{\"id\": \"facebook\",\"identificationType\": \"facebook\",\"enabled\": true}," +
				"{\"id\": \"twitter\",\"identificationType\": \"twitter\",\"enabled\": true}," +
				"{\"id\": \"instagram\",\"identificationType\": \"instagram\",\"enabled\": true}," +
				"{\"id\": \"tigold\",\"identificationType\": \"tigold\",\"enabled\": true}," +
				"{\"id\": \"national-electorial-court\",\"identificationType\": \"national-electorial-court\",\"enabled\": true}," +
				"{\"id\": \"other\",\"identificationType\": \"other\",\"enabled\": true}]}"
		};
		user = { attributes: { firstName: "Keke", lastName: "Roseberg" } } as User;

		const singleTermCustomers = SimpleDataMock.getSingleTermCustomersMock();

		const location: RouteComponentProps<any>["location"] = {
			hash: "",
			key: "",
			pathname: "",
			search: "",
			state: {}
		};
		const history: RouteComponentProps<any>["history"] = {
			length: 2,
			action: "POP",
			location: location,
			push: jest.fn(),
			replace: jest.fn(),
			go: jest.fn(),
			goBack: jest.fn(),
			goForward: jest.fn(),
			block: jest.fn(),
			createHref: jest.fn(),
			listen: jest.fn()
		};
		const routerProps: RouteComponentProps<any> = {
			match: {
				isExact: true,
				params: {},
				path: "",
				url: ""
			},
			location: location,
			history: history,
			staticContext: {}
		};

		minProps = {
			...routerProps,
			user: user,
			searchConfigs: consulStore.searchConfigs,
			singleTerm: "Timo Isokoski",
			singleTermCustomers: singleTermCustomers,
			searchingForSingleCustomer: true,
			hasActiveCustomerCase: true,
			actions: {
				clearSingleTermSearch: (...args: any[]) => {},
				searchCustomerWithSingleTerm: (...args: any[]) => {},
				setSingleTerm: (...args: any[]) => {},
				showCustomerCreationModal: (...args: any[]) => {},
				setCustomer: (...args: any[]) => {},
				createNewCustomerCase: (...args: any[]) => {},
				deleteUIbasket: (...args: any[]) => {},
				cancelAddProduct: (...args: any[]) => {},
				resetConfigurations: (...args: any[]) => {},
				resetAddressWithBasketItemIdEntries: () => {},
				resetLocations: () => {},
			}
		};

		store = TestUtils.mockReduxStore({
			currency: {
				currencies: [],
				selectedCurrency: ""
			},
			basket: {},
			consul: consulStore,
			customerCase: {
				activeCustomerCase: {}
			},
			navBar: {
				showLanguageSelect: true,
				showBasket: true
			},
			user: {
				user: user
			},
			feature: {}
		});

		context = {
			flux: {
				stores: {
					BasketStore: makeStore("context.flux.stores.BasketStore"),
					ConsulStore: makeStore("context.flux.stores.ConsulStore", consulStore),
					CurrencyStore: makeStore("context.flux.stores.CurrencyStore"),
					SalesRepSessionStore: makeStore("context.flux.stores.SalesRepSessionStore"),
					UserStore: makeStore("context.flux.stores.UserStore", { user: user })
				},
				actions: {
					BasketActions: {},
					UserActions: {
						aaLogin: () => {}
					}
				}
			},
			store: store
		};
	});

	it("succeeds at shallow mount with minimum props", () => {
		const wrapper = shallowWithContext(<CustomerSelection {...minProps} />);
		expect(wrapper).toMatchSnapshot();
	});

	it("succeeds at deep mount with minimum props", () => {
		mountWithContext(<CustomerSelection {...minProps} />, { context });
	});

	it("renders with minimum props", () => {
		// const { getValueAndTextFromSelectOptions } = TestUtils;

		const wrapper: ReactWrapper = mountWithContext(<CustomerSelection {...minProps} />, {
			context
		});

		expect(
			wrapper
				.find("#navbar-search")
				.hostNodes()
				.prop("value")
		).toEqual(minProps.singleTerm);
	});

	/* Phone number came with RND-11721 */
	/* Config from consul with additional fields came with RND-213617
	/* apps/omnichannel-pos/1/properties/features/pos_search_configuration */
	it("presents Customer, Phone number, Passport, Identification-Card and Identification-Number as Search criteria", () => {
		const wrapper = mountWithContext(<CustomerSelection {...minProps} />, {
			context
		});
		wrapper.find("#navbar_select_menu").hostNodes().simulate("click");

		const searchCriteriaDropdown = wrapper.find("#navbar_select_menu_dropdown").hostNodes();

		expect(searchCriteriaDropdown
				.find("button")
				.map((node: any) => node.text())
				.sort()
		).toEqual(
			[
				"Customer",
				"Diplomatic Carnet",
				"Disability Cell",
				"Driver's Licence",
				"Facebook",
				"Foreign Document",
				"ID-Card",
				"Instagram",
				"Military ID-card",
				"Minor ID-card",
				"National Electorial Court",
				"Other",
				"Passport",
				"Subscription Phone Number",
				"Residence Number",
				"Student ID",
				"Tax ID",
				"Temporary Work Permit",
				"Tigold",
				"Twitter"
			].sort()
		);
	});

	it("allows searching for a Customer by Phone number", done => {
		let wrapper: ReactWrapper;
		let props: CustomerSelectionProps;

		const { findNodesWithOnClickAttribute } = TestUtils;

		const searchTerm = "1239487";

		const CustomerStore = {
			singleTermCustomers: [],
			searchingForSingleCustomer: false,
			singleTerm: ""
		};

		props = {
			...minProps,
			singleTermCustomers: CustomerStore.singleTermCustomers,
			searchingForSingleCustomer: CustomerStore.searchingForSingleCustomer,
			singleTerm: CustomerStore.singleTerm
		};
		props.actions.setSingleTerm = (value: string) => {
			if (wrapper && props) {
				props.singleTerm = value;
				wrapper.setProps(props);
			}
		};

		props.actions.searchCustomerWithSingleTerm = (value: string, criteria: string) => {
			expect(value).toEqual(searchTerm);
			expect(criteria).toEqual("number");
			done();
		};

		context.flux.CustomerStore = CustomerStore;

		wrapper = mountWithContext(<CustomerSelection {...props} />, {
			context
		});
		wrapper.find("#navbar_select_menu").hostNodes().simulate("click");

		const searchCriteriaDropdown = wrapper.find("#navbar_select_menu_dropdown");
		const phoneNumberOption = searchCriteriaDropdown
			.find("button")
			.filterWhere((n: any) => n.text().toLowerCase() === "Subscription Phone Number".toLowerCase());
		findNodesWithOnClickAttribute(phoneNumberOption).simulate("click");

		const searchForm = wrapper.find("form");
		searchForm.find("#navbar-search").simulate("change", { target: { value: searchTerm } });
		wrapper.update();
		searchForm.find("#navbar_search_button").find("button").simulate("click");
	});

	it("passes search term to CustomerListDropdown", done => {
		let props: CustomerSelectionProps;
		let wrapper: ReactWrapper;

		const searchTerm = "Timo Isokoski";
		const actions: CustomerSelectionProps["actions"] = {
			...minProps.actions,
			setSingleTerm: (value: string) => {
				if (props && wrapper) {
					props.singleTerm = value;
					wrapper.setProps(props);
				}
			},
			searchCustomerWithSingleTerm: (value: string, criteria: string) => {
				expect(value).toEqual(searchTerm);
				expect(criteria).toEqual("name");
				done();
			}
		};

		props = {
			...minProps,
			singleTerm: searchTerm,
			actions
		};

		wrapper = mountWithContext(<CustomerSelection {...props} />, {
			context
		});

		const searchForm = wrapper.find("#w-app-search");
		searchForm.find("#navbar-search").simulate("change", { target: { value: searchTerm } });
		wrapper.update();
		searchForm.find("#navbar_search_button").find("button").simulate("click");
		wrapper.update();
		expect((wrapper.find("CustomerListDropdown").props() as CustomerListDropdownProps).searchTerm).toEqual(
			searchTerm
		);
	});

	describe("RUBT-58895", () => {
		let newContext: any;
		let wrapper: ReactWrapper;
		let props: CustomerSelectionProps;

		const singleTermCustomers = SimpleDataMock.getSingleTermCustomersMock();

		store = TestUtils.mockReduxStore({
			currency: {
				currencies: [],
				selectedCurrency: ""
			},
			basket: {},
			consul: consulStore,
			customer: {
				singleTermCustomers,
				searchingForSingleCustomer: true,
				singleTerm: singleTermCustomers[0].attributes.formattedName
			},
			customerCase: {
				activeCustomerCase: {}
			},
			navBar: {
				showLanguageSelect: true,
				showBasket: true
			},
			user: {
				salesRepUser: undefined,
				user: { firstName: "Keke", lastName: "Roseberg" } as User
			},
			feature: {}
		});

		beforeEach(() => {
			newContext = {
				...context,
				store,
				flux: {
					store: {
						CustomerStore: {
							singleTermCustomers,
							searchingForSingleCustomer: true,
							singleTerm: singleTermCustomers[0].attributes.formattedName
						},
						CustomerCaseStore: {
							activeCustomerCase: {}
						},
						UserStore: {
							salesRepUser: null,
							user: { firstName: "Keke", lastName: "Roseberg" }
						},
						ConsulStore: consulStore
					},
					actions: {
						UserActions: {
							aaLogin: () => {}
						}
					}
				}
			};
			const actions: CustomerSelectionProps["actions"] = {
				...minProps.actions,
				setSingleTerm: (value: string) => {
					if (!value && wrapper && props) {
						wrapper.setProps({
							...props,
							singleTermCustomers: [],
							searchingForSingleCustomer: false,
							singleTerm: ""
						});
					}
				},
				searchCustomerWithSingleTerm: (value: string, criteria: string) => {
					window.console.log(value);
					// 	expect(value).toEqual(searchTerm);
					// 	expect(criteria).toEqual("number");
					// 	done();
				}
			};

			props = {
				...minProps,
				actions: actions,
				singleTermCustomers,
				searchingForSingleCustomer: true,
				singleTerm: singleTermCustomers[0].attributes.formattedName,
				user: { firstName: "Keke", lastName: "Roseberg" } as User
			};
		});

		afterEach(() => {
			wrapper.unmount();
		});

		/* RUBT-58895 */
		it("must not clear search terms no matter how many times the search is executed", () => {
			const { getRandomIntBetweenZeroAndN } = TestUtils;

			wrapper = mountWithContext(<CustomerSelection {...props} />, { context: newContext });

			const minimumTrialsCount = 3;
			const trials = getRandomIntBetweenZeroAndN(4) + minimumTrialsCount;

			const searchForm = wrapper.find("form");
			const searchInput = wrapper.find("#navbar-search").hostNodes();
			searchInput.simulate("change", {
				target: {
					value: singleTermCustomers[0].attributes.formattedName
				}
			});

			const doSearch = (trialNumber: string) => {
				const propValue = searchInput.prop("value");
				const expectedValue = singleTermCustomers[0].attributes.formattedName;

				expect(propValue).toEqual(expectedValue);

				searchForm.simulate("submit");

				const results = wrapper.find("CustomerListDropdown");
				const customerList = results.find(".CustomerListDropdown-customer-list-dropdown");
				const customerItems = customerList
					.find(".CustomerListDropdown-dropdown-body")
					.filterWhere((node: any) => !node.hasClass("CustomerListDropdown-tile-wrapper"));

				expect(customerItems.length).toBeGreaterThan(0);
			};
			for (let i = 1; i <= trials; ++i) {
				doSearch(`${i}`);
			}
		});

		/* RUBT-58895 */
		it.skip("should clear search results and terms when customer is chosen", () => {
			const { findNodesWithOnClickAttribute } = TestUtils;

			wrapper = mountWithContextAndRouterProps(<CustomerSelection {...props} />, {
				context: newContext
			});

			const searchForm = wrapper.find("form");
			const searchInput = wrapper.find("#navbar-search").hostNodes();
			searchInput.simulate("change", {
				target: {
					value: singleTermCustomers[0].attributes.formattedName
				}
			});
			searchForm.simulate("submit");

			const results = wrapper.find("CustomerListDropdown");
			let customerList = results.find(".CustomerListDropdown-customer-list-dropdown");
			const customerItems0 = customerList
				.find(".CustomerListDropdown-tile-wrapper")
				.filterWhere(node => !node.hasClass("CustomerListDropdown-customer-dropdown-header"));
			expect(customerItems0.length).toBeGreaterThan(0);

			findNodesWithOnClickAttribute(customerItems0)
				.hostNodes()
				.simulate("click");
			customerList = wrapper.find("CustomerListDropdown").find(".CustomerListDropdown-customer-list-dropdown");
			const customerItems1 = customerList
				.find(".CustomerListDropdown-tile-wrapper")
				.hostNodes()
				.filterWhere(node => !node.hasClass("CustomerListDropdown-customer-dropdown-header"));
			expect(customerItems1.length).toEqual(0);
		});

		/* RUBT-58895 */
		it("hides search results for the time that focus is outside the search elements", () => {
			wrapper = mountWithContext(<CustomerSelection {...props} />, { context: newContext });

			const searchForm = wrapper.find("form");
			const searchInput = wrapper.find("#navbar-search").hostNodes();
			searchInput.simulate("change", {
				target: {
					value: singleTermCustomers[0].attributes.formattedName
				}
			});
			searchForm.simulate("submit");

			const results0 = wrapper.find("CustomerListDropdown");
			const customerList0 = results0.find(".CustomerListDropdown-customer-list-dropdown");
			const customerItems0 = customerList0
				.find(".CustomerListDropdown-dropdown-body")
				.filterWhere(node => !node.hasClass("CustomerListDropdown-tile-wrapper"));
			expect(customerItems0.length).toBeGreaterThan(0);

			TestUtils.clickOnDocumentBody(() => {
				(results0.instance() as any).handleClickOutside();
			});
			wrapper.update();
			const results1 = wrapper.find("CustomerListDropdown");
			const customerList1 = results1.find(".CustomerListDropdown-customer-list-dropdown");
			expect(customerList1.length).toEqual(0);

			wrapper.find("form").simulate("submit");
			wrapper.update();

			const results2 = wrapper.find("CustomerListDropdown");
			const customerList2 = results2.find(".CustomerListDropdown-customer-list-dropdown");

			const customerItems2 = customerList2
				.find(".CustomerListDropdown-dropdown-body")
				.filterWhere(node => !node.hasClass("CustomerListDropdown-tile-wrapper"));
			expect(customerItems2.length).toBeGreaterThan(0);
		});
	});
});
