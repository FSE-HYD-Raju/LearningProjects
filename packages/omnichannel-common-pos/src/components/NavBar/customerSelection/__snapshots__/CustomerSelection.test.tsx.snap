// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerSelection succeeds at shallow mount with minimum props 1`] = `
<form
  className="NavBar-customer-selection"
  id="w-app-search"
  onSubmit={[Function]}
>
  <div
    className="input-group"
  >
    <Connect(IntlProvider)>
      <OnClickOutside(NavBarSearchMenu)
        criteria="name"
        eventTypes={
          Array [
            "mousedown",
            "touchstart",
          ]
        }
        excludeScrollbar={false}
        handleCriteriaChange={[Function]}
        outsideClickIgnoreClass="ignore-react-onclickoutside"
        preventDefault={false}
        stopPropagation={false}
        unparsedSearchConfigs="{\\"POSSearchConfigs\\":[{\\"id\\": \\"name\\", \\"enabled\\": true},{\\"id\\": \\"phone-number\\",\\"enabled\\": true},{\\"id\\": \\"passport\\",\\"identificationType\\": \\"passport\\",\\"enabled\\": true},{\\"id\\": \\"minor-ID\\",\\"identificationType\\": \\"minor-ID\\",\\"enabled\\": true},{\\"id\\": \\"ID-card\\",\\"identificationType\\": \\"identification-card\\",\\"enabled\\": true},{\\"id\\": \\"military-ID\\",\\"identificationType\\": \\"military-ID\\",\\"enabled\\": true},{\\"id\\": \\"foreign-document\\",\\"identificationType\\": \\"foreign-document\\",\\"enabled\\": true},{\\"id\\": \\"diplomatic-carnet\\",\\"identificationType\\": \\"diplomatic-carnet\\",\\"enabled\\": true},{\\"id\\": \\"student-ID\\",\\"identificationType\\": \\"student-ID\\",\\"enabled\\": true},{\\"id\\": \\"tax-ID\\",\\"identificationType\\": \\"tax-ID\\",\\"enabled\\": true},{\\"id\\": \\"drivers-licence\\",\\"identificationType\\": \\"drivers-licence\\",\\"enabled\\": true},{\\"id\\": \\"residence-number\\",\\"identificationType\\": \\"residence-number\\",\\"enabled\\": true},{\\"id\\": \\"work-permit\\",\\"identificationType\\": \\"work-permit\\",\\"enabled\\": true},{\\"id\\": \\"disability-cell\\",\\"identificationType\\": \\"disability-cell\\",\\"enabled\\": true},{\\"id\\": \\"facebook\\",\\"identificationType\\": \\"facebook\\",\\"enabled\\": true},{\\"id\\": \\"twitter\\",\\"identificationType\\": \\"twitter\\",\\"enabled\\": true},{\\"id\\": \\"instagram\\",\\"identificationType\\": \\"instagram\\",\\"enabled\\": true},{\\"id\\": \\"tigold\\",\\"identificationType\\": \\"tigold\\",\\"enabled\\": true},{\\"id\\": \\"national-electorial-court\\",\\"identificationType\\": \\"national-electorial-court\\",\\"enabled\\": true},{\\"id\\": \\"other\\",\\"identificationType\\": \\"other\\",\\"enabled\\": true}]}"
      />
    </Connect(IntlProvider)>
    <input
      className="form-control"
      id="navbar-search"
      onChange={[Function]}
      value="Timo Isokoski"
    />
    <div
      className="input-group-append"
    >
      <OcButton
        buttonType="DEFAULT"
        htmlBtnType="submit"
        id="navbar_search_button"
        onClick={[Function]}
      >
        <i
          className="fa fa-search"
        />
      </OcButton>
    </div>
  </div>
</form>
`;
